### PR TITLE
tmpfiles: resolve specifiers from the target root

### PR DESCRIPTION
**Pull request by nolange**
the --root option is supposed to set up an isolated,
dormant system.
For that its necessary to inspect the target root for
information that is necessary for machine-id, hostname
and similar specifiers.

Some open points are what should happen with `--root=/` and what if `--user` is specified as well (should that even be possible?)

Fixes: #16183

### CODE REVIEW

File: src/tmpfiles/tmpfiles.c [hunk 1]
```diff
@@ -178,20 +181,25 @@ STATIC_DESTRUCTOR_REGISTER(arg_include_prefixes, freep);
...
-        { 'b', specifier_boot_id,         NULL },
-        { 'H', specifier_host_name,       NULL },
-        { 'l', specifier_short_host_name, NULL },
-        { 'v', specifier_kernel_release,  NULL },
-        { 'a', specifier_architecture,    NULL },
-        { 'o', specifier_os_id,           NULL },
-        { 'w', specifier_os_version_id,   NULL },
-        { 'B', specifier_os_build_id,     NULL },
-        { 'W', specifier_os_variant_id,   NULL },
+        { 'b', specifier_boot_id_isol,    NULL },
```
Comment (poettering) [inline src/tmpfiles/tmpfiles.c]: isol? that sounds like a weird suffix. What's that even supposed to mean? short for "isolated"?

File: src/tmpfiles/tmpfiles.c [hunk 2]
```diff
@@ -208,20 +216,148 @@ static const Specifier specifier_table[] = {
...
+        if (sd_id128_is_null(machine_id))
+                return -ENOMEDIUM;
+
+        n = new (char, 33);
+        if (!n)
+                return -ENOMEM;
+
+        *ret = sd_id128_to_string(machine_id, n);
+        return r;
+}
```
Comment (poettering) [inline src/tmpfiles/tmpfiles.c]: hmm, so this will come up in other tools too. I think it would be better to just teach the generic specifier_machine_id() the trick instead of adding this to every tool individually. I.e. use the `data` argument to specifier_machine_id() (which we currently pass as NULL always) to pass in the root dir to operate on.

File: src/tmpfiles/tmpfiles.c [hunk 3]
```diff
@@ -208,20 +216,148 @@ static const Specifier specifier_table[] = {
...
+        return id128_read(fullpath, f, ret);
+}
+
+static int specifier_machine_id_isol(
+        const char *prefix, char specifier, const void *data, const void *userdata, char **ret) {
+        int r;
+        char *n;
+        sd_id128_t machine_id = {};
+
+        r = id128_read_isol(prefix, "/etc/machine-id", ID128_PLAIN, &machine_id);
```
Comment (poettering) [inline src/tmpfiles/tmpfiles.c]: this must use chase_symlinks() so that this follows symlinks relative to the specified root.

Also if arg_root is not set this really should be shortcut to use sd_id128_get_machine(), as before, since that is cached and so on...

File: src/tmpfiles/tmpfiles.c [hunk 4]
```diff
@@ -208,20 +216,148 @@ static const Specifier specifier_table[] = {
...
+}
+
+static int specifier_arch_isol(char specifier, const void *data, const void *userdata, char **ret) {
+        if (!empty_or_root(arg_root))
+                return -ENXIO;
+
+        return specifier_architecture(specifier, data, userdata, ret);
+}
+
+static int id128_read_isol(const char *prefix, const char *p, Id128Format f, sd_id128_t *ret) {
```
Comment (poettering) [inline src/tmpfiles/tmpfiles.c]: prefix is not actually used here...

File: src/tmpfiles/tmpfiles.c [hunk 5]
```diff
@@ -208,20 +216,148 @@ static const Specifier specifier_table[] = {
...
+static int specifier_kernel_isol(char specifier, const void *data, const void *userdata, char **ret) {
+        if (!empty_or_root(arg_root))
+                return -ENXIO;
+
+        return specifier_kernel_release(specifier, data, userdata, ret);
+}
+
+static int specifier_arch_isol(char specifier, const void *data, const void *userdata, char **ret) {
+        if (!empty_or_root(arg_root))
+                return -ENXIO;
```
Comment (poettering) [inline src/tmpfiles/tmpfiles.c]: I think these two should always work the same way. i.e. i think it's something to document in tmpfiles.d man page instead of prohibiting it altogether... i.e. we should just say `%a always resolves to the host's architecture, regardless if '--root=' is used...`  or so

**Comment by poettering** [comment]
> Some open points are what should happen with `--root=/` and what if `--user` is specified as well (should that even be possible?)

I'd probably just prohibit the combination... Better prohibit too much...

File: src/tmpfiles/tmpfiles.c (hunk 5 above)
Comment (nolange) [reply src/tmpfiles/tmpfiles.c]: Depends on what `--root=` is trying to archive. I would like to build a root filesystem for another architecture and pre-create all files & folders (outside of volatile paths like /tmp and /run). If its the same arch I could execute `systemd-tmpfiles` in a chroot.
Needless to say I would rather have the specifiers missing than wrong. I guess you are mostly thinking about containers for the same machine?
What is `--root=` used for currently?

File: src/tmpfiles/tmpfiles.c (hunk 1 above)
Comment (nolange) [reply src/tmpfiles/tmpfiles.c]: Yes, suggest a better name ? `rooted`?

File: src/tmpfiles/tmpfiles.c (hunk 5 above)
Comment (boucman) [reply src/tmpfiles/tmpfiles.c]: maybe use an environment variable or a separate command line argument to override it ?

Comment (poettering) [reply src/tmpfiles/tmpfiles.c]: @nolange but the point I am making is that your patch doesn't make it work either, it just causes the stuff to fail.

I think @boucman has a point: let's maybe make this overridable via an env var, and independent of arg_root or no arg_root. i.e. if `$SYSTEMD_ARCHITECTURE=` is set, then %a would resolve to that rather than the actual architecture. Similar for the other fields.

Comment (poettering) [reply src/tmpfiles/tmpfiles.c]: > What is --root= used for currently?

Mostly testing I guess: i.e. run some stuff without affecting the system things are run on.

I mean, I sympathize with the goal of this patch, and I am convinced for the specifiers that are sourced from files of the OS tree we should also operate relative to the same root specified by `--root=`, but this doesn't really work so well for %a and the like, that expose runtime properties more than actual properties of the directory tree. For those I think a better approach is to just inherit from the host, and maybe make this overridable via env var.

Comment (nolange) [reply src/tmpfiles/tmpfiles.c]: > @nolange but the point I am making is that your patch doesn't make it work either, it just causes the stuff to fail.

**Decisively** fail, just like there should be no boot-uuid. Its well-defined that the entries containing those specifiers will be skipped, those entries will be created when the relevant specifiers are available ("real" boot).

In the issue report  #16183 I noted the option to add overrides, and this is the path I would like to take.
The safe default however would be leave out specifiers that are not unambiguously available.

Comment (poettering) [reply src/tmpfiles/tmpfiles.c]: I really wouldn't blur the line between runtime parameters (i.e. %a, %b, %H, %v, …) and properties of the disk image itself (i.e. %B, %w, %m, …). They are quite different in nature, the former are derived from all kinds of things, but the latter only from the image itself. I wouldn't try to be too smart here, beacuse the runtime parameters are only very very indirectly connected to the disk image, hence I would say that anything is good enough for them, and if people want explicit control over runtime parameters when doing "offline" operation, then env vars are the way to go for those. I don#t think there's value in prohibiting using the current system's runtime params for --root= operation however, because for our test cases that would break stuff...

I mean, you never know what people use --root= for. it could be that they do "staged" stuff, i.e. apply things into a subdir to the copy/merge it into the host fs. In such a case it is totally OK if the runtime specifiers resolve like they normally do.

Hence again: the arg_root stuff should affect %B, %w, %m, absolutely, happy to merge a patch for that. But for %a, %b %H I think a concept of making them overridable via env vars is the better approach is it permits folks who want to set this up for full offline operation to do what they want, but it doesn't limit anything for anyone else.

Comment (nolange) [reply src/tmpfiles/tmpfiles.c]: > In such a case it is totally OK if the runtime specifiers resolve like they normally do.

Or you might accidentally leak some information from your host, if you stage a build for a different machine. I'd stay on the side of caution, but I definitely would need **one** switch to block all current and future specifiers.

Ill plan to do the following the next days:

- change the default to pick runtime params from the host
- add a switch `--disable-runtime-specifiers` to change the default to how I (first) implemented it: return -ENXIO, skip affected lines.
- add argument `--set-specifier={boot-id,hostname,architecture,kernel-release}=VALUE` for replacing the values.
  potentially need some special `%host%` and `%null%` VALUES to only switch few values.

Comment (poettering) [reply src/tmpfiles/tmpfiles.c]: i think controlling this via env vars makes sense, but I'd prefer not to have high-level switches. "override" style stuff we generally do via env vars, most options in docs/ENVIRONMENT.md are override style stuff. I think we should do this the same way here. There's no shame in using env vars like that.

also, if possible don't do "negative" options, i.e. "--disable-foo"... turning something off by turning it on it not the prettiest thing I believe. This is only a request though, and the threshold for a solution is low. for example "mask" or "prohibit" or "refuse" might already be enough to use instead of "disable" to make me shut up about this. Better though is making it positive. 

Maybe do:

```
$SYSTEMD_SPECIFIER_BOOT_ID=…
$SYSTEMD_SPECIFIER_HOSTNAME=…
$SYSTEMD_SPECIFIER_PERMIT_RUNTIME=<bool>
…
```

File: src/tmpfiles/tmpfiles.c (hunk 1 above)
Comment (nolange) [reply src/tmpfiles/tmpfiles.c]: changed in new version

File: src/tmpfiles/tmpfiles.c (hunk 3 above)
Comment (nolange) [reply src/tmpfiles/tmpfiles.c]: now using chase_symlinks_and_open

File: src/tmpfiles/tmpfiles.c (hunk 2 above)
Comment (nolange) [reply src/tmpfiles/tmpfiles.c]: extended a function in `id128-util.c`, and the `specifier_machine_id` function

File: src/tmpfiles/tmpfiles.c (hunk 4 above)
Comment (nolange) [reply src/tmpfiles/tmpfiles.c]: function was removed

File: src/tmpfiles/tmpfiles.c (hunk 5 above)
Comment (nolange) [reply src/tmpfiles/tmpfiles.c]: I think I leave this open, as this is the main thread of discussion, I am not done yet and I will go over the formatting later. Got little spare time now, just tell me if I am on the right track so far.

The tool now works identical as before, until you set `SYSTEMD_SPECIFIER_PROHIBIT_RUNTIME`, which is intended to
-   do the same thing as the `systemd-tmpfiles-setup.service` on a `--root=prefix`. But take into account that this might be a different machine than the host
-   specifiers that cant be unambiguous resolved should cause the entry to be skipped (need to thing of an better warning msg there).
-   some `SYSTEMD_SPECIFIER_` env variables should allow setting a few specifiers directly (TODO)

**Comment by mrc0mmand** [comment]
CI no likey:

```
[    8.272520] Run /init as init process
[    8.277138]   with arguments:
[    8.280898]     /init
[    8.283881]   with environment:
[    8.287845]     HOME=/
[    8.291103]     TERM=linux
[    8.294562]     raid=noautodetect
[    8.298771]     SYSTEMD_UNIT_PATH=/usr/lib/systemd/tests/testdata/testsuite-01.units:/usr/lib/systemd/tests/testdata/units:
systemd 246-rc1-4-ga39cfee running in system mode. (+PAM +AUDIT -SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +ZSTD +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=unified)
Virtualization QEMU found in DMI (/sys/class/dmi/id/sys_vendor)
Virtualization found, CPUID=KVMKVMKVM
Found VM virtualization kvm
Detected virtualization kvm.
Detected architecture x86-64.
Running in initial RAM disk.
Failed to read os-release file, ignoring: No such file or directory

Welcome to Linux!

Assertion 'fd >= 0' failed at src/libsystemd/sd-id128/id128-util.c:84, function id128_read_fd(). Aborting.
Caught <ABRT>, dumped core as pid 213.
Freezing execution.
qemu-system-x86_64: terminating on signal 15 from pid 78153 (timeout)
E: test timed out after 600 s
```

**Comment by nolange** [comment]
CI is good boy.
Fixed

**Comment by nolange** [comment]
Ok, clened up formatting, adding a description to `docs/ENVIRONMENT.md` and added a check to disallow `--root` and `--user` at the same time.

For me this is sufficient, I might add some more ENV variables to override Hostname ad a later time.

**Comment by nolange** [comment]
Ping

File: src/libsystemd/sd-id128/id128-util.c [hunk 6]
```diff
@@ -128,12 +128,18 @@ int id128_read_fd(int fd, Id128Format f, sd_id128_t *ret) {
...
 
-int id128_read(const char *p, Id128Format f, sd_id128_t *ret) {
+int id128_read_prefix(const char *p, Id128Format f, sd_id128_t *ret, const char *root) {
         _cleanup_close_ int fd = -1;
 
-        fd = open(p, O_RDONLY|O_CLOEXEC|O_NOCTTY);
-        if (fd < 0)
-                return -errno;
+        if (root) {
+                fd = chase_symlinks_and_open(p, root, CHASE_PREFIX_ROOT, O_RDONLY | O_CLOEXEC, NULL);
```
Comment (poettering) [inline src/libsystemd/sd-id128/id128-util.c]: doesn't matter much, but hy drop the O_NOCTTY here? maybe some weirdo points us to some tty here...

File: src/libsystemd/sd-id128/id128-util.c [hunk 7]
```diff
@@ -128,12 +128,18 @@ int id128_read_fd(int fd, Id128Format f, sd_id128_t *ret) {
         return sd_id128_from_string(buffer, ret);
 }
 
-int id128_read(const char *p, Id128Format f, sd_id128_t *ret) {
+int id128_read_prefix(const char *p, Id128Format f, sd_id128_t *ret, const char *root) {
```
Comment (poettering) [inline src/libsystemd/sd-id128/id128-util.c]: please keep return params last, i.e. swap the two last args, or maybe even put the root arg second

File: src/shared/specifier.c [hunk 8]
```diff
@@ -219,19 +224,19 @@ static int specifier_os_release_common(const char *field, char **ret) {
 }
 
 int specifier_os_id(char specifier, const void *data, const void *userdata, char **ret) {
-        return specifier_os_release_common("ID", ret);
+        return specifier_os_release_common((const char *) data, "ID", ret);
```
Comment (poettering) [inline src/shared/specifier.c]: the cast should not be needed?

File: src/shared/specifier.h [hunk 9]
```diff
@@ -6,8 +6,8 @@
 typedef int (*SpecifierCallback)(char specifier, const void *data, const void *userdata, char **ret);
 
 typedef struct Specifier {
-        const char specifier;
-        const SpecifierCallback lookup;
+        char specifier;
+        SpecifierCallback lookup;
```
Comment (poettering) [inline src/shared/specifier.h:10]: hmm, split this bit into its own commit pls

File: src/tmpfiles/tmpfiles.c [hunk 10]
```diff
@@ -181,33 +191,81 @@ STATIC_DESTRUCTOR_REGISTER(arg_root, freep);
...
+        { { 'U', specifier_user_id },         SPECTABLE_USERDB },
+        { { 'u', specifier_user_name },       SPECTABLE_USERDB },
+        { { 'h', specifier_user_home },       SPECTABLE_USERDB },
+
+        { { 't', specifier_directory, UINT_TO_PTR(DIRECTORY_RUNTIME) }, SPECTABLE_DIRECTORY },
+        { { 'S', specifier_directory, UINT_TO_PTR(DIRECTORY_STATE) },   SPECTABLE_DIRECTORY },
+        { { 'C', specifier_directory, UINT_TO_PTR(DIRECTORY_CACHE) },   SPECTABLE_DIRECTORY },
+        { { 'L', specifier_directory, UINT_TO_PTR(DIRECTORY_LOGS) },    SPECTABLE_DIRECTORY },
+        { { 'T', specifier_tmp_dir,     "/tmp" },     SPECTABLE_STRING },
+        { { 'V', specifier_var_tmp_dir, "/var/tmp" }, SPECTABLE_STRING },
```
Comment (poettering) [inline src/tmpfiles/tmpfiles.c]: not sure i follow here. Why not just define two `Specifier[]` tables, one for the off and one for the on case?

File: src/tmpfiles/tmpfiles.c [hunk 11]
```diff
@@ -181,33 +191,81 @@ STATIC_DESTRUCTOR_REGISTER(arg_root, freep);
...
+                                break;
+                        case SPECTABLE_RUNTIME:
+                        case SPECTABLE_USERDB:
+                                spec.lookup = disabled_specifier;
+                                break;
+                        default:
+                                break;
+                        }
+                } else {
+                        // remove arguments only used for offline
```
Comment (poettering) [inline src/tmpfiles/tmpfiles.c]: as per coding style, we use `//` comments for `FIXME` style stuff only, and `/* */` for explanatory comments.

File: src/libsystemd/sd-id128/id128-util.c (hunk 6 above)
Comment (nolange) [reply src/libsystemd/sd-id128/id128-util.c]: No particular reason (that I can remember). Added back the flag

File: src/shared/specifier.c (hunk 8 above)
Comment (nolange) [reply src/shared/specifier.c]: I changed the argument for `specifier_os_release_common`, so its needed bc of that

File: src/tmpfiles/tmpfiles.c (hunk 10 above)
Comment (nolange) [reply src/tmpfiles/tmpfiles.c]: The idea is to allow en/disabling groups of specifiers, and overwriting of specifiers via ENV variables. I ran into limitations with setting the hostname, as there are 2 related specifiers.
I am all ears if you can propose a good scheme.

File: src/libsystemd/sd-id128/id128-util.c [hunk 12]
```diff
@@ -128,12 +128,18 @@ int id128_read_fd(int fd, Id128Format f, sd_id128_t *ret) {
...
-                return -errno;
+        if (root) {
+                fd = chase_symlinks_and_open(p, root, CHASE_PREFIX_ROOT, O_RDONLY | O_CLOEXEC | O_NOCTTY, NULL);
+                if (fd < 0)
+                        return fd;
+        } else {
+                fd = open(p, O_RDONLY | O_CLOEXEC | O_NOCTTY);
+                if (fd < 0)
+                        return -errno;
+        }
```
Comment (poettering) [inline src/libsystemd/sd-id128/id128-util.c]: you can drop the second branch. chase_symlinks_and_open() already shortcuts things like this if root is NULL or "/"

File: src/shared/specifier.c [hunk 13]
```diff
@@ -120,7 +122,10 @@ int specifier_machine_id(char specifier, const void *data, const void *userdata,
         char *n;
         int r;
 
-        r = sd_id128_get_machine(&id);
+        if (empty_or_root((const char *) data))
+                r = sd_id128_get_machine(&id);
+        else
+                r = id128_read_prefix("/etc/machine-id", (const char *) data, ID128_PLAIN, &id);
```
Comment (poettering) [inline src/shared/specifier.c]: the cast is unnecessary, `const void*` implicitly casts to all other `const` pointer types

File: src/shared/specifier.c [hunk 14]
```diff
@@ -219,19 +224,19 @@ static int specifier_os_release_common(const char *field, char **ret) {
 }
 
 int specifier_os_id(char specifier, const void *data, const void *userdata, char **ret) {
-        return specifier_os_release_common("ID", ret);
+        return specifier_os_release_common((const char *) data, "ID", ret);
 }
```
Comment (poettering) [inline src/shared/specifier.c:228]: the cast is not necessary still. as mentioned above "const void*" casts implicitly to "const char*" and the other way too!

File: src/tmpfiles/tmpfiles.c (hunk 10 above)
Comment (poettering) [reply src/tmpfiles/tmpfiles.c]: i don't get why you need these groups, though?

Comment (nolange) [reply src/tmpfiles/tmpfiles.c]: First, I have the values from the default table and need to replace parts of it, so I already use this info (see table). Of course right now I could just use 2 tables, but...
But take for example the os-release or hostname specifiers, I planned that those could have 3-4 different states:

-   enabled (pick from running system)
-   disabled (skip lines containing those specs)
-   from file in root prefix
-   optionally overridden explicitly

grouping the specs should allow stuff like changing long + short hostname.

The other solution I considered is running all logic at start and building a table with `specifier_string` entries. Would be more straight up.

Comment (poettering) [reply src/tmpfiles/tmpfiles.c]: i really don#t see the usecase for the complex modes. even if there's need for them eventually I'd start out simple and just do "on" and "off"

File: src/shared/specifier.c (hunk 13 above)
Comment (nolange) [reply src/shared/specifier.c]: Ok, C++ guy here. Seems like something a static analysis tool might get mad about

File: src/libsystemd/sd-id128/id128-util.c (hunk 12 above)
Comment (nolange) [reply src/libsystemd/sd-id128/id128-util.c]: ok

File: src/tmpfiles/tmpfiles.c (hunk 10 above)
Comment (nolange) [reply src/tmpfiles/tmpfiles.c]: Well, maybe you see now with the new version. I cant just add the prefix to a static table, I just don't see the simple solution you seem to have in mind.

File: src/tmpfiles/tmpfiles.c [hunk 15]
```diff
@@ -212,6 +214,29 @@ static const Specifier specifier_table[] = {
...
+        for (i = 0; i < (int) ELEMENTSOF(specifier_table) - 1; ++i) {
+                const char spec = specifier_table[i].specifier;
+                if (memchr("mowBW", spec, sizeof("mowBW") - 1))
+                        /* specifier takes an argument for root prefix */
+                        specifier_table[i].data = prefix;
+                else if (spec == 'T' || spec == 'V') {
+                        specifier_table[i].lookup = specifier_string;
+                        specifier_table[i].data = spec == 'T' ? "/tmp" : "/var/tmp";
+                } else if (specifier_table[i].lookup != specifier_directory)
+                        specifier_table[i].lookup = disabled_specifier;
```
Comment (poettering) [inline src/tmpfiles/tmpfiles.c:236]: instead of patching around in the specifier table, make them constant please, but have two around, one for the no-runtime case and one for the normal case. Yes, they will be mostly idnetical, but that's fine.

**Review by poettering — CHANGES_REQUESTED** [review]

File: src/tmpfiles/tmpfiles.c (hunk 15 above)
Comment (nolange) [reply src/tmpfiles/tmpfiles.c:236]: I can't make them constant, unless I use some static array for `prefix`. Otherwise the pointer to prefix will only be known at runtime

Comment (poettering) [reply src/tmpfiles/tmpfiles.c:236]: specifier_printf() has a third param, `userdata`, we only pass `NULL` there so far. Pass the `prefix` there instead.

**Comment by aduskett** [comment]
Any update on this? It would be an excellent addition for a patch series that's been sitting for a while in Buildroot
https://patchwork.ozlabs.org/project/buildroot/patch/20200615072055.2083-11-nolange79@gmail.com/

If the specifier_printf() issue is all that's left, I would be more than happy to submit those changes!

**Comment by keszybz** [comment]
> If the specifier_printf() issue is all that's left, I would be more than happy to submit those changes!

Please do. There's no guarantee that other things will not pop up in further review, this is open source after all, but at the moment it seems that this series is almost ready.

**Comment by poettering** [comment]
ping?

**Comment by aduskett** [comment]
> ping?

I'll try to work on the last remaining issues this week. The last few weeks have been crazy busy! Sorry about that!

**Comment by nolange** [comment]
In case the ping was at me, I wont touch this for a while.

Primary reason because I believe shoehorning this into `userdata` will mean that this will be part of the interface and would need some extensibility. You then cant use specifier that expects something else than a rootfs prefix in `userdata`, at least at the same time.
So there would need to be some extensible struct passed in.

I tried to keep the changes local to the `tmpfiles` tool, not re-think the "specifiers" framework.

**Comment by poettering** [comment]
> In case the ping was at me, I wont touch this for a while.

The PR looks so close to being ready though?

> Primary reason because I believe shoehorning this into `userdata` will mean that this will be part of the interface and would need some extensibility. You then cant use specifier that expects something else than a rootfs prefix in `userdata`, at least at the same time.
> So there would need to be some extensible struct passed in.
> 
> I tried to keep the changes local to the `tmpfiles` tool, not re-think the "specifiers" framework.

No need to. The "userdata" pointer is not used in tmpfiles at all so far, all calls to specifier_printf() in tmpfiles just pass it as NULL right now. The parameter exists only for cases like this, use it! 

You shouldn't need to touch the generic specifier handling code at all, all changes can remain local to tmpfiles.d. All that needs to be done is replace some specifier handler functions listed in the `specifier_table[]` and passing the prefix in the so far unused userdata argument.

**Comment by nolange** [comment]
> > In case the ping was at me, I wont touch this for a while.
> 
> The PR looks so close to being ready though?

Not sure.
>
> > Primary reason because I believe shoehorning this into `userdata` will mean that this will be part of the interface and would need some extensibility. You then cant use specifier that expects something else than a rootfs prefix in `userdata`, at least at the same time.
> > So there would need to be some extensible struct passed in.
> > I tried to keep the changes local to the `tmpfiles` tool, not re-think the "specifiers" framework.
> 
> No need to. The "userdata" pointer is not used in tmpfiles at all so far, all calls to specifier_printf() in tmpfiles just pass it as NULL right now. The parameter exists only for cases like this, use it!

If I change, say `specifier_os_version_id` so that it will interpret `userdata` as prefix, then funky things will happen if
it is used elsewhere, with say a `Unit *` 
https://github.com/systemd/systemd/blob/427103f7a627af9a89b5eb470b13c8ff50ca43ae/src/core/unit-printf.c#L227

> 
> You shouldn't need to touch the generic specifier handling code at all, all changes can remain local to tmpfiles.d. All that needs to be done is replace some specifier handler functions listed in the `specifier_table[]` and passing the prefix in the so far unused userdata argument.

I cant change  the specifier callbacks to use the userdata. Or do you imply that I wrap every specifier callback?

I would expect that a separate table is alot less hassle, and one thats not `const` could be patched if one wants to add further overrides in the future.

**Comment by poettering** [comment]
I posted PR #20017 now. It's an attempt to solve this thoroughly across the tree, i.e. also in sysusers, repart, systemctl. It doesn't do the blocking of the resolution of the runtime specifiers, because it's not clear to me this is good and necessary.

Let's close this one and continue there.
